### PR TITLE
Pybind11 fix linkage

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **pybind11**

* Update to Conan V2 and new toolchains
* Attempt to fix issue raised in #11075

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by planetmarshall** [comment]
Something is causing pybind to link in `libpython` which it shouldn't be doing for a module. A modification to the test package can workaround the issue, but I'd still like to get to the bottom of it as it will break consumers using the typical path.

```py
# pybind11 method:
pybind11_add_module(MyModule1 src1.cpp)
```

Results in a linking error because `libpython` is linked in 

```py
# Python method:
Python_add_library(MyModule2 src2.cpp)
target_link_libraries(MyModule2 pybind11::headers)
```

However, links correctly.

It's a bit mysterious as conan targets *only* links the headers, they don't do anything else. Linking in Python (or not) is handled by pbind11's custom CMake scripts.

See [Pybind11 CMake helpers](https://pybind11.readthedocs.io/en/latest/cmake/index.html#suggested-usage)

File: recipes/pybind11/all/test_package/conanfile.py [hunk 1]
```diff
@@ -21,7 +42,7 @@ def _python_interpreter(self):
...
 
     def test(self):
-        if not tools.cross_building(self):
-            with tools.environment_append({"PYTHONPATH": "lib"}):
-                self.run("{} {}".format(self._python_interpreter, os.path.join(
-                    self.source_folder, "test.py")), run_environment=True)
+        if can_run(self):
+            python_path = os.path.join(self.build_folder, self.cpp.build.libdirs[0])
+            module_path = os.path.join(self.source_folder, "test.py")
+            self.run(f"PYTHONPATH={python_path} {self._python_interpreter} {module_path}", env="conanrun")
```
Comment (planetmarshall) [inline recipes/pybind11/all/test_package/conanfile.py:48]: Is there a better way to set this in Conan V2? I can't set it in `generate` as I don't have access to `self.build_folder` at that point.

**Review by jwillikers — DISMISSED** [review]

File: recipes/pybind11/all/test_package/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/pybind11/all/test_package/conanfile.py]: That's a great question. Hopefully someone who knows will update us.

File: recipes/pybind11/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -1,16 +1,9 @@
...
-conan_set_fpic()
-conan_check_compiler()
-conan_set_libcxx()
-conan_set_vs_runtime()
-
+find_package(Python COMPONENTS Interpreter Development)
 find_package(pybind11 REQUIRED CONFIG)
-
-pybind11_add_module(test_package MODULE test_package.cpp)
+#pybind11_add_module(test_package MODULE test_package.cpp)
```
[the remark this answers is not in the document]
Comment (planetmarshall) [reply recipes/pybind11/all/test_package/CMakeLists.txt:6]: PR still in draft - there are still a few things I want to test.

**Comment by planetmarshall** [comment]
fixed the component dependency order - the main changes are

* set the main target name to `pybind11_do_not_use` (similar to cairo and other packages) so that a dummy main component can be created
* create a dummy main component called `pybind11::pybind11`
* `pybind11::pybind11` should depend on `pybind11::headers`, and not the other way around.

**Review by jwillikers — DISMISSED** [review]

File: recipes/pybind11/all/conanfile.py [hunk 3]
```diff
@@ -16,83 +15,63 @@ class PyBind11Conan(ConanFile):
...
-        if self.settings.compiler == "apple-clang" and Version(self.settings.compiler.version) >= "11.0":
-            raise ConanInvalidConfiguration("OSX support is bugged. Check https://github.com/pybind/pybind11/issues/3081")
+    def generate(self):
+        toolchain = CMakeToolchain(self)
+        toolchain.cache_variables.update({
+            "PYBIND11_INSTALL": True,
+            "PYBIND11_TEST": False,
+            "PYBIND11_CMAKECONFIG_INSTALL_DIR": "lib/cmake/pybind11"
+        })
+        toolchain.generate()
```
Comment (prince-chrismc) [inline recipes/pybind11/all/conanfile.py:34]: use cache variables sparingly please

``suggestion
        tc = CMakeToolchain(self)
        tc.variables["PYBIND11_INSTALL"] = True
        tc.variables["PYBIND11_TEST"] = False
        tc.variables["PYBIND11_CMAKECONFIG_INSTALL_DIR"] = "lib/cmake/pybind11"
        tc.generate()
```

File: recipes/pybind11/all/conanfile.py [hunk 4]
```diff
@@ -16,83 +15,63 @@ class PyBind11Conan(ConanFile):
...
+        self.cpp_info.components["pybind11_"].set_property("cmake_target_name", "pybind11::pybind11")
+        self.cpp_info.components["pybind11_"].set_property("cmake_module_file_name", "pybind11")
+        self.cpp_info.components["pybind11_"].names["cmake_find_package"] = "pybind11"
+        self.cpp_info.components["pybind11_"].builddirs = [cmake_base_path]
+        self.cpp_info.components["pybind11_"].requires = ["headers"]
+        cmake_file = os.path.join(cmake_base_path, "pybind11Common.cmake")
+        self.cpp_info.set_property("cmake_build_modules", [cmake_file])
+        for generator in ["cmake_find_package", "cmake_find_package_multi"]:
+            self.cpp_info.components["_pybind11"].build_modules[generator].append(cmake_file)
+        self.cpp_info.components["embed"].requires = ["_pybind11"]
```
Comment (prince-chrismc) [inline recipes/pybind11/all/conanfile.py:70]: Above the component is `pybind11_` which does it change?

Comment (planetmarshall) [reply recipes/pybind11/all/conanfile.py]: typo

**Review by jwillikers — DISMISSED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Comment by ericLemanissier** [comment]
@planetmarshall please close the PR for 10 second then reopen it so that it retriggers a build and finally merge this change

**Review by jwillikers — APPROVED** [review]

**Comment by uilianries** [comment]
My guess is, it failed with a runtime error. Need to run on Windows to check it.

**Review by uilianries — APPROVED** [review]

**Pull request merged**